### PR TITLE
fix web_seed redirects

### PR DESCRIPTION
**Pull request by ajax16384**
- make all redirects ephemeral (to avoid resume serialization)
- remove dup slash for redirects in proxy mode

### CODE REVIEW

File: src/web_peer_connection.cpp [hunk 1]
```diff
@@ -411,7 +411,16 @@ void web_peer_connection::write_request(peer_request const& r)
 			auto redirection = m_web->redirects.find(f.file_index);
 			if (redirection != m_web->redirects.end())
 			{
-				request += redirection->second;
+				auto const& redirect = redirection->second;
+				if (using_proxy && !redirect.empty() && (redirect[0] == '/'))
```
Comment (arvidn) [inline src/web_peer_connection.cpp]: a more concise way of expressing this would be:

```
bool const trailing_slash = using_proxy && !redirect.empty() && redirect[0] == '/';
request.append(redirect, trailing_slash, std::string::npos);
```

Comment (arvidn) [inline src/web_peer_connection.cpp]: why is ``using_proxy`` a condition here btw?

File: src/http_seed_connection.cpp [hunk 2]
```diff
@@ -315,7 +315,7 @@ namespace libtorrent
 					}
 
 					// add the redirected url and remove the current one
-					t->add_web_seed(location, web_seed_entry::http_seed);
+					t->add_web_seed(location, web_seed_entry::http_seed, std::string(), web_seed_entry::headers_t(), true);
```
Comment (arvidn) [inline src/http_seed_connection.cpp]: I take it that last ``true`` means ephemeral, right? why do you need that here?

Comment (arvidn) [inline src/http_seed_connection.cpp]: iirc, it just means it won't be saved in the resume data, but you still remove the original web seed URL below, won't that be a problem?

**Comment by arvidn** [comment]
could you think of a way to test this patch? I believe there's a simulation you could extend to make sure you don't get duplicate ``/``

File: src/web_peer_connection.cpp (hunk 1 above)
Comment (ajax16384) [inline src/web_peer_connection.cpp]: "m_url" will contain "http://example.com/" and in http proxy case will be added to "request" variable. if "redirection->second" contains "/redirect_file" then "request" may contain "http://example.com//redirect_file".

File: src/http_seed_connection.cpp (hunk 2 above)
Comment (ajax16384) [inline src/http_seed_connection.cpp]: redirected web_seed should not be stored in resume because it is only serialize url part (without "redirects" map") e.g original url points to "http://example1.com/file1", web server response with location "http://example2.com/subpath/file2". (m_url=="http://example2.com/", redirects[0]=="/subpath/file2"). if we try to load old save resume then library tryies to request wrong path "http://example2.com/file1" ("redirects" map is lost).  alternative approach is to serialize "redirects" maps, but it will definetly increase resume data size.

File: src/web_peer_connection.cpp (hunk 1 above)
Comment (ajax16384) [inline src/web_peer_connection.cpp]: "trailing_slash" - ok

**Comment by ajax16384** [comment]
- test\web_server.py:http_handler should validate income request url (dissalow duplicate slashes after "protocol://" and raise exception in case of rule violation
- test\test_web_seed_redirect.cpp should be extended with web_seed_redirect_through_proxy

File: src/web_peer_connection.cpp (hunk 1 above)
Comment (arvidn) [inline src/web_peer_connection.cpp]: but even if ``using_proxy`` is false, concatenating those two paths may result in a double slash, no?

File: src/http_seed_connection.cpp (hunk 2 above)
Comment (arvidn) [inline src/http_seed_connection.cpp]: what do you mean by that? done in serial?

File: src/web_peer_connection.cpp [hunk 3]
```diff
@@ -640,7 +643,7 @@ void web_peer_connection::handle_redirect(int const bytes_left)
 
 		// add_web_seed won't add duplicates. If we have already added an entry
 		// with this URL, we'll get back the existing entry
-		web_seed_t* web = t->add_web_seed(redirect_base, web_seed_entry::url_seed, m_external_auth, m_extra_headers);
+		web_seed_t* web = t->add_web_seed(redirect_base, web_seed_entry::url_seed, m_external_auth, m_extra_headers, true);
```
Comment (arvidn) [inline src/web_peer_connection.cpp]: could you please explain why these two web seeds should be ephemeral? and confirm that all it means is that they won't be saved in the resume data. (ideally as a brief comment in the code)

**Comment by arvidn** [comment]
if ``web_server.py`` checks for double slashes, how come the tests aren't failing without this patch?
In general it would be better to add tests as simulations, as the python web server (and socks proxy) isn't as reliable and I'm trying to move away from having those kinds of tests

File: src/web_peer_connection.cpp (hunk 3 above)
Comment (ajax16384) [inline src/web_peer_connection.cpp]: redirected web_seed should not be stored in resume because it is only serialize url part (without "redirects" map") e.g original url points to "http://example1.com/file1", web server response with location "http://example2.com/subpath/file2". (m_url=="http://example2.com/", redirects[0]=="/subpath/file2"). if we try to load old save resume then library tryies to request wrong path "http://example2.com/file1" ("redirects" map is lost).

Comment (ajax16384) [inline src/web_peer_connection.cpp]: brief comment should be placed at "bool ephemeral" declaration?

**Comment by ajax16384** [comment]
if web_server.py will find duplicate slash then it should throw exception that fail/abort current unit test.

**Comment by arvidn** [comment]
> if web_server.py will find duplicate slash then it should throw exception that fail/abort current unit test.

Right. If it would and a test would trigger the case that causes a double slash, that would cover this bug. none of those conditions are true right now though.

> simulations: you talking about adding checks to simulation\test_web_seed.cpp ?

yes. in fact, the check is implicit, since you register handlers for specific paths. if the path is wrong your handle won't get run. so it should just be a matter of triggering the double slash in a test.

File: src/web_peer_connection.cpp (hunk 3 above)
Comment (arvidn) [inline src/web_peer_connection.cpp]: I would think a brief comment where you add the new (redirected) urls would make sense

Comment (arvidn) [inline src/web_peer_connection.cpp]: at some point I will change that ``bool`` parameter to be a ``enum class ephemeral { persist, dont_persist };``, to make the call site readable.

**Comment by ajax16384** [comment]
upon implementation of web_seed redirect through proxy simulation i have found test framework error.  look at following pseudo workflow:
````c++
create_dir('tmp123');
change_dir('tmp123');
  test_a();
    storage.add_file('foo1');
    try_to_download_generated_content_to('foo1');
    wait_till_seed_mode();
  test_b();
    storage.add_file('foo1'); // same file name as test_a()
    try_to_download_generated_content_to('foo1'); // we have already file from test_a(), so libTorrent reuse it's content without simulation network access, ANY possible network logic error checks will be passed
    wait_till_seed_mode(); 
 ````
to fix such behaviour i fixed test/main.cpp to add sub directory for every unit tests. this will help isolate unit test file access.

**Comment by ajax16384** [comment]
Next error was at libsimulator (wrong memmove usage) which was fixed at https://github.com/arvidn/libsimulator/pull/22

**Comment by ajax16384** [comment]
@arvidn now proxy simulator falls at:
Assertion failed: false, file libsimulator\src\acceptor.cpp, line 175
````c++
tcp::acceptor::incoming_packet(p) , p.type == aux::packet::payload
````
to repro just apply libsimulator commit, this commit and uncomment: https://github.com/arvidn/libtorrent/pull/1651/commits/75e239d6d539d3db0f70f6cf9c817c88f0d83902#diff-154d5e15c0604c37b9a2b4316318441bR432

File: test/main.cpp [hunk 4]
```diff
@@ -357,6 +368,17 @@ EXPORT int main(int argc, char const* argv[])
 		if (filter && tests_to_run.count(_g_unit_tests[i].name) == 0)
 			continue;
 
+		std::snprintf(dir, sizeof(dir), "sub_%i", i);
+		std::string unit_dir = combine_path(current_working_directory(), dir);
```
Comment (arvidn) [inline test/main.cpp]: this looks questionable. you overwrite "dir" here (as far as I can tell) with the per-test sub directory, the you add that to the current working directory, then you set the new current working directory to that.

I just notice that you "cd .." at the end of the test. What would you think about saving the root test directory you generate above, and reuse that to append the per-test subdirectory to. that way each way through the loop you would just cd straight to the new subdirectory.

File: simulation/test_web_seed.cpp [hunk 5]
```diff
@@ -404,6 +406,67 @@ TORRENT_TEST(multi_file_redirect)
...
+		{
+			settings_pack pack;
+
+			pack.set_int(settings_pack::proxy_type, settings_pack::http);
+			pack.set_str(settings_pack::proxy_hostname, "50.50.50.50");
+			pack.set_str(settings_pack::proxy_username, "testuser");
+			pack.set_str(settings_pack::proxy_password, "testpass");
+			pack.set_int(settings_pack::proxy_port, 4445);
+			pack.set_bool(settings_pack::proxy_hostnames, true);
+			//TODO: ses.apply_settings(pack);
```
Comment (arvidn) [inline simulation/test_web_seed.cpp]: thanks! I hope to get a chance to check this out this weekend

File: test/main.cpp (hunk 4 above)
Comment (arvidn) [inline test/main.cpp]: in fact, since this is in master (and you have C++11) you could just do:

auto unit_dir = combine_path(dir, "sub_" + std::to_string(i));

**Comment by arvidn** [comment]
did you see this btw? https://travis-ci.org/arvidn/libtorrent/jobs/198212458#L1706

File: test/main.cpp (hunk 4 above)
Comment (ajax16384) [inline test/main.cpp]: ok

**Comment by ajax16384** [comment]
yes, will fix in next commit

File: src/file.cpp [hunk 6]
```diff
@@ -503,6 +503,29 @@ namespace libtorrent
 #endif
 	}
 
+	void change_directory(std::string const& f, error_code& ec)
```
Comment (arvidn) [inline src/file.cpp]: I would actually prefer if this function stays in the test util library. libtorrent should never update current working directory. But it does make sense for the tests

File: test/main.cpp [hunk 7]
```diff
@@ -368,16 +338,21 @@ EXPORT int main(int argc, char const* argv[])
...
-			std::printf("Failed to create unit directory: %s\n", ec.message().c_str());
+			std::printf("Failed to create unit test directory: %s\n", ec.message().c_str());
 			return 1;
 		}
-		if (!change_dir(dir)) return 1; // every unit test must have separate sandbox
+		change_directory(unit_dir, ec);
+		if (ec)
+		{
+			std::printf("Failed to change unit test directory: %s\n", ec.message().c_str());
+			return 1;
```
Comment (arvidn) [inline test/main.cpp]: could this (and the return above) be a ``continue`` instead?

File: test/main.cpp [hunk 8]
```diff
@@ -368,16 +338,21 @@ EXPORT int main(int argc, char const* argv[])
...
+		}
+		unit_dirs.push_back(unit_dir);
```
Comment (arvidn) [inline test/main.cpp]: the directory should be put in here immediately after it's been (successfully) created. imagine you can create the directory, but fail to cd into it. you still want to delete it later.

In fact, I'm kind of thinking that it would be more robust to delete the test directory inside the loop, right after each individual test passes. Would that make sense you think? Then you wouldn't need to store them in a vector.

In addition, the way I think they should be deleted is by having a scope guard delete it (so it happens regardless of early exits, continues and exceptions). Something like this:
``struct dir_guard { std::string dir; ~dir_guard() { error_code ec; remove_all(dir, ec); } } guard = { unit_dir };``

Comment (arvidn) [inline test/main.cpp]: I'd be happy to make this change after yours is in though. because I also don't understand why the directories are only cleaned up if ``TORRENT_LOGGING`` is not defined. that doesn't seem to make any sense

Comment (ajax16384) [inline test/main.cpp]: ok

File: src/file.cpp (hunk 6 above)
Comment (ajax16384) [inline src/file.cpp]: ok

File: test/main.cpp (hunk 7 above)
Comment (ajax16384) [inline test/main.cpp]: i think it is unrecoverable fail state of test framework (some base functionality that must prepare environment for unit test). if it cannot change dir then there is no sense to continue.

**Comment by ajax16384** [comment]
@arvidn if you decrease file sizes of old "unaligned_file_redirect" test so both files will fall in one piece, then it will fail.

File: test/main.cpp (hunk 7 above)
Comment (arvidn) [inline test/main.cpp]: true, it's critical the process exits with an error code if this happens

File: src/web_peer_connection.cpp [hunk 9]
```diff
@@ -649,14 +658,39 @@ void web_peer_connection::handle_redirect(int const bytes_left)
...
+			{
+				std::string redirect_path_prefix = redirect_path.substr(0, redirect_path_idx);
+				file_storage const& fs = info.files();
+				for (file_index_t fi(0); fi < fs.end_file(); ++fi)
+				{
+					if (info.orig_files().pad_file_at(fi))
+						continue;
+					if (web->redirects.find(fi) != web->redirects.end())
+						continue;
+					web->have_files.set_bit(fi);
```
Comment (arvidn) [inline src/web_peer_connection.cpp]: why are you doing this? it seems incorrect. you're assuming that all files will be redirected to this server. It seems to defeat the purpose of the ``have_files`` bitmask, which records which files have *actually* ben redirected here.

Comment (arvidn) [inline src/web_peer_connection.cpp]: btw, is this patch a prerequisite for your other fix? If this change is independent, I would prefer it be a separate PR

File: src/web_peer_connection.cpp [hunk 10]
```diff
@@ -780,9 +814,8 @@ void web_peer_connection::on_receive(error_code const& error
 				peer_log(peer_log_alert::info, "STATUS"
 					, "%d %s", m_parser.status_code(), m_parser.message().c_str());
 				std::multimap<std::string, std::string> const& headers = m_parser.headers();
-				for (std::multimap<std::string, std::string>::const_iterator i = headers.begin()
-					, end(headers.end()); i != end; ++i)
-					peer_log(peer_log_alert::info, "STATUS", "   %s: %s", i->first.c_str(), i->second.c_str());
+				for (auto const &i : headers)
```
Comment (arvidn) [inline src/web_peer_connection.cpp]: +1

File: src/web_peer_connection.cpp [hunk 11]
```diff
@@ -380,11 +385,8 @@ void web_peer_connection::write_request(peer_request const& r)
 		std::vector<file_slice> files = info.orig_files().map_block(req.piece, req.start
 			, req.length);
 
-		for (std::vector<file_slice>::iterator i = files.begin();
-			i != files.end(); ++i)
+		for (auto const &f : files)
```
Comment (arvidn) [inline src/web_peer_connection.cpp]: +1

File: test/main.cpp [hunk 12]
```diff
@@ -217,6 +219,51 @@ void print_usage(char const* executable)
...
+#endif
+}
+
+struct unit_directory_guard
+{
+	std::string dir;
+	~unit_directory_guard()
+	{
+		error_code ec;
+		change_directory(root_dir, ec); // windows will not allow to remove current dir, so let's change it to root
```
Comment (arvidn) [inline test/main.cpp]: instead of making ``root_dir`` a global variable, could you just do:

``change_directory(parent_path(dir));``?

File: test/main.cpp [hunk 13]
```diff
@@ -217,6 +219,51 @@ void print_usage(char const* executable)
...
+		if (ec)
+		{
+			std::printf("Failed to change directory: %s\n", ec.message().c_str());
```
Comment (arvidn) [inline test/main.cpp]: this should probably be a fatal error. You should be able to use the ``TEST_ERROR`` macro for that, I would imagine.

File: test/main.cpp [hunk 14]
```diff
@@ -217,6 +219,51 @@ void print_usage(char const* executable)
...
+			return;
+		}
+		if (_g_test_failures == 0 && !keep_files)
```
Comment (arvidn) [inline test/main.cpp]: would you mind always deleting the files unless ``keep_files`` is set?

File: test/test.cpp [hunk 15]
```diff
@@ -36,7 +36,8 @@ POSSIBILITY OF SUCH DAMAGE.
 
 unit_test_t _g_unit_tests[1024];
 int _g_num_unit_tests = 0;
-int _g_test_failures = 0;
+int _g_test_failures = 0; // flushed at start of every unit
+int _g_test_total_failures = 0;
```
Comment (arvidn) [inline test/test.cpp]: did this used to return incorrect return code from the test process? if the last test passed, would it shadow previous failures?
Is there a good reason to clear ``_g_test_failures`` each turn through the loop? I suppose there needs to be some way to tell whether a test had failures or not. an alternative way would be to copy this counter at the beginning of the test and then compare at the end. I think that would be a bit nicer, as it would require one less global variable.

**Comment by arvidn** [comment]
btw, I merged your PR against libsimulator, so I imagine you could update the submodule reference in this patch

File: src/web_peer_connection.cpp (hunk 9 above)
Comment (ajax16384) [inline src/web_peer_connection.cpp]: this is optimisitic assumption that all the rest of files will be at these new predicted locations (it's done only once if redirects[file_index] is not filled). if assumption is incorrect then have_file.clear_bit occured for this web_seed.
anyway it's the current only way to download unaligned files from plain redirect (new location for all files are equal by file_path string endings). without this prediction download simply stuck cause web_peer_connection::on_connect cut requests for edge pieces.

File: test/main.cpp (hunk 12 above)
Comment (ajax16384) [inline test/main.cpp]: ok

File: test/main.cpp (hunk 13 above)
Comment (ajax16384) [inline test/main.cpp]: ok

File: test/main.cpp (hunk 14 above)
Comment (ajax16384) [inline test/main.cpp]: ok

File: test/test.cpp (hunk 15 above)
Comment (ajax16384) [inline test/test.cpp]: you right, it shadowed real failures count by last successfull result.
test framework already had some total_failures counter so i just moved it to global scope.

File: src/web_peer_connection.cpp (hunk 9 above)
Comment (ajax16384) [inline src/web_peer_connection.cpp]: with optimistic prediction web_seed requests will be issued to new location at first and only if it failures to origin server.
just as i said the real problem may be observed even at old "unaligned_file_redirect" with several files which content resides inside first piece.

**Comment by ajax16384** [comment]
ok, but i still waiting for how to pass libsimulator\src\acceptor.cpp assert

File: test/test.cpp (hunk 15 above)
Comment (ajax16384) [inline test/test.cpp]: and print_failures() is called in the end of all test and must use real total number of failures which may be located only at it's file definition (test/test.cpp)

**Comment by arvidn** [comment]
oh, right

**Comment by arvidn** [comment]
I believe the problem with the simulation is that the http proxy only supports a single request at a time. It was made this way to keep it simple, and it covered all the use cases at the time.

This test specifically want to redirect files to separate servers though, right? It can't be changed to only have a single proxy connection at a time?

File: src/web_peer_connection.cpp (hunk 9 above)
Comment (arvidn) [inline src/web_peer_connection.cpp]: that case is supposed to be handled by storing partial pieces in the ``web_seed_entry``. I suppose it's a problem when you combine that with redirects, since you have multiple different ``web_seed_entry`` objects. I would prefer to extend the existing solution than to do this though. It seems a bit hacky, and doesn't solve the general problem (afaict).

File: test/test.cpp (hunk 15 above)
Comment (arvidn) [inline test/test.cpp]: actually, aren't the number of failures stored per unit test in the ``_g_unit_tests`` array?
I would expect the ``test_total_failures`` state is redundant, and could be recomputed from the test array. I haven't looked into this recently, but I suspect there's a smaller/simpler patch to solve the problem. Btw, what was the symptom you experienced?

**Comment by ajax16384** [comment]
unfortunately no

File: test/test.cpp (hunk 15 above)
Comment (ajax16384) [inline test/test.cpp]: oh. you right - it can be recalculated

Comment (ajax16384) [inline test/test.cpp]: symptom - unexpected test return code

**Comment by arvidn** [comment]
I'm still working on libsimulator to get this to work

**Comment by arvidn** [comment]
I've fixed libsimulator here https://github.com/arvidn/libsimulator/pull/24
You could test with that branch, but as soon as it passes CI, I will merge it into master

**Comment by arvidn** [comment]
that fix is in master of libsimulator now

**Comment by arvidn** [comment]
I'm tempted to pull out your fixes to the test framework and back-port them to RC_1_1.

**Comment by ajax16384** [comment]
libsimulator commit fix proxy unit test, now it web_seed redirect through proxy work as intended. (with commented out dup slash checking unit test throw error)

**Comment by ajax16384** [comment]
@arvidn i can backport only test framework changes by myself

**Comment by arvidn** [comment]
cool. I looked into it last night, and as far as I can tell, there's nothing wrong with counting or reporting the number of errors. However, your patch to have each individual test run in its own directory would probably improve reliability of the tests

**Comment by ajax16384** [comment]
https://github.com/arvidn/libtorrent/pull/1707

**Comment by ajax16384** [comment]
waiting for another rebase round

**Comment by arvidn** [comment]
please rebase on top of master again now

File: test/test.cpp [hunk 16]
```diff
@@ -65,7 +62,7 @@ int print_failures()
 	}
 
 	std::printf("\n\n");
-	int total_num_failures = 0;
+	int total_num_failures { 0 };
```
Comment (arvidn) [inline test/test.cpp]: is there a good reason to make this change?

File: test/test.cpp [hunk 17]
```diff
@@ -51,7 +49,6 @@ void report_failure(char const* err, char const* file, int line)
 	char buf[500];
 	std::snprintf(buf, sizeof(buf), "\x1b[41m***** %s:%d \"%s\" *****\x1b[0m\n", file, line, err);
 	std::fprintf(stderr, "\n%s\n", buf);
-	failure_strings.push_back(buf);
```
Comment (arvidn) [inline test/test.cpp]: I used to print all of these at the end. I take it that's not the case anymore? I think it's a pretty convenient feature, to potentially avoid digging through a bunch of log output

File: test/main.cpp [hunk 18]
```diff
@@ -514,6 +507,8 @@ EXPORT int main(int argc, char const* argv[])
 	if (redirect_stdout) fflush(stdout);
 	if (redirect_stderr) fflush(stderr);
 
-	return print_failures() ? 333 : 0;
+	int total_num_failures = print_failures();
+
+	return total_num_failures ? 333 : 0;
```
Comment (arvidn) [inline test/main.cpp]: is there a good reason for this change? If so, ``total_num_failures`` should be const

File: test/main.cpp [hunk 19]
```diff
@@ -361,10 +359,8 @@ EXPORT int main(int argc, char const* argv[])
 #else
 	process_id = getpid();
 #endif
-	std::string root_dir = current_working_directory();
-	char dir[40];
-	snprintf(dir, sizeof(dir), "test_tmp_%u", process_id);
-	std::string unit_dir_prefix = combine_path(root_dir, dir);
+	std::string root_dir { current_working_directory() };
```
Comment (arvidn) [inline test/main.cpp]: if anything, this should be made const. I don't see a strong reason to use uniform initialization here, instead of regular initialization.

File: test/main.cpp [hunk 20]
```diff
@@ -221,20 +221,18 @@ void print_usage(char const* executable)
 void change_directory(std::string const& f, error_code& ec)
 {
 	ec.clear();
+	native_path_string n = convert_to_native_path_string(f);
```
Comment (arvidn) [inline test/main.cpp]: I believe this ``n`` can be ``const``, right?

File: src/web_peer_connection.cpp (hunk 9 above)
Comment (arvidn) [inline src/web_peer_connection.cpp]: I would prefer to not do this. A better approach would be to factor out the "partial piece" from ``web_seed_entry`` into the torrent object, or something at the torrent level. keyd by the piece. That way, a partial piece downloaded from one web seed can be picked up an resume by another

File: simulation/test_web_seed.cpp [hunk 21]
```diff
@@ -383,6 +386,118 @@ TORRENT_TEST(multi_file_redirect)
...
+		{
+			settings_pack pack;
+
+			pack.set_int(settings_pack::proxy_type, settings_pack::http);
+			pack.set_str(settings_pack::proxy_hostname, "50.50.50.50");
+			pack.set_str(settings_pack::proxy_username, "testuser");
+			pack.set_str(settings_pack::proxy_password, "testpass");
+			pack.set_int(settings_pack::proxy_port, 4445);
+			pack.set_bool(settings_pack::proxy_hostnames, true);
+			ses.apply_settings(pack);
```
Comment (arvidn) [inline simulation/test_web_seed.cpp]: I see you enabled this now. has the submodule been updated to the latest version already?

File: src/web_peer_connection.cpp [hunk 22]
```diff
@@ -126,6 +122,15 @@ web_peer_connection::web_peer_connection(peer_connection_args const& pack
 #endif
 }
 
+std::string web_peer_connection::escape_file_path(file_storage const& storage, file_index_t index)
```
Comment (arvidn) [inline src/web_peer_connection.cpp]: as far as I can tell, this does not need to be a member function. I think it should be a free function with internal linkage (anonymous namespace). and its declaration should be removed from the header

File: test/test.cpp (hunk 16 above)
Comment (ajax16384) [inline test/test.cpp]: no, just remained from old commits, will revert to minmize patch

File: test/test.cpp (hunk 17 above)
Comment (ajax16384) [inline test/test.cpp]: ok, will revert

File: test/main.cpp (hunk 18 above)
Comment (ajax16384) [inline test/main.cpp]: ok

File: simulation/test_web_seed.cpp (hunk 21 above)
Comment (ajax16384) [inline simulation/test_web_seed.cpp]: yes

File: test/main.cpp (hunk 20 above)
Comment (ajax16384) [inline test/main.cpp]: ok

**Comment by ajax16384** [comment]
@arvidn build failed due: https://github.com/arvidn/libsimulator/pull/25

**Comment by ajax16384** [comment]
if cwd returned in UNC format then many tests failed. reverting from current commit. will make separate PR after this one.

**Comment by arvidn** [comment]
I'm not sure I understand the cwd point. But internally in libtorrent, I believe the assumption is that paths are utf8 in posix format. is that right?

**Comment by ajax16384** [comment]
i meant case then current_working_directory() returns something like "\\?\c:\abc\def"

File: src/web_peer_connection.cpp (hunk 9 above)
Comment (arvidn) [inline src/web_peer_connection.cpp]: I still consider this a blocker for this PR btw.

Comment (ajax16384) [inline src/web_peer_connection.cpp]: ok

**Comment by arvidn** [comment]
lgtm
